File: apps/mobile/test/core/logic/confirm_trade_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:fabtrades/core/logic/confirm_trade.dart';
import 'package:fabtrades/core/models/binder_entry.dart';
import 'package:fabtrades/core/models/trade.dart';

import '../../support/fixtures.dart';

void main() {
  final now = DateTime.utc(2026, 7, 1);

  BinderEntry binder(String id, int qty, {bool wanted = false}) => BinderEntry(
        card: buildCard(id: id),
        quantity: qty,
        isWanted: wanted,
        addedAt: now,
      );

  group('reconcileBinderAfterTrade', () {
    test('decrements given cards and clamps at zero', () {
      final trade = Trade(
        id: 't',
        createdAt: now,
        haveItems: [
          TradeItem(card: buildCard(id: 'a'), quantity: 3, priceEach: 1),
        ],
      );
      final next = reconcileBinderAfterTrade(
        entries: [binder('a', 1)],
        trade: trade,
        removeGivenFromBinder: true,
        addReceivedToBinder: false,
        now: now,
      );
      expect(next, isEmpty);
    });

    test('adds received cards to binder with NM merge', () {
      final card = buildCard(id: 'b');
      final trade = Trade(
        id: 't',
        createdAt: now,
        wantItems: [TradeItem(card: card, quantity: 2, priceEach: 1)],
      );
      final next = reconcileBinderAfterTrade(
        entries: [binder('b', 1)],
        trade: trade,
        removeGivenFromBinder: false,
        addReceivedToBinder: true,
        now: now,
      );
      expect(next.single.quantity, 3);
      expect(next.single.condition, 'NM');
      expect(next.single.isWanted, isFalse);
    });

    test('clears received cards from want list even when add is off', () {
      final card = buildCard(id: 'c');
      final trade = Trade(
        id: 't',
        createdAt: now,
        wantItems: [TradeItem(card: card, quantity: 1, priceEach: 1)],
      );
      final next = reconcileBinderAfterTrade(
        entries: [binder('c', 2, wanted: true)],
        trade: trade,
        removeGivenFromBinder: false,
        addReceivedToBinder: false,
        now: now,
      );
      expect(next.single.quantity, 1);
      expect(next.single.isWanted, isTrue);
    });

    test('skips binder mutations when both checkboxes are off', () {
      final given = buildCard(id: 'g');
      final recv = buildCard(id: 'r');
      final trade = Trade(
        id: 't',
        createdAt: now,
        haveItems: [TradeItem(card: given, quantity: 1, priceEach: 1)],
        wantItems: [TradeItem(card: recv, quantity: 1, priceEach: 1)],
      );
      final start = [binder('g', 5), binder('r', 1, wanted: true)];
      final next = reconcileBinderAfterTrade(
        entries: start,
        trade: trade,
        removeGivenFromBinder: false,
        addReceivedToBinder: false,
        now: now,
      );
      // Given stays; received want still cleared.
      expect(next.where((e) => e.card.id == 'g').single.quantity, 5);
      expect(next.where((e) => e.card.id == 'r'), isEmpty);
    });
  });
}

File: apps/web/tests/utils/confirmTrade.test.js
import {
    reconcileBinderAfterTrade,
    tradeLinesFromList,
} from '../../src/utils/confirmTrade.js';

const binder = (cardId, quantity, { wanted = false, condition = 'NM' } = {}) => ({
    cardId,
    quantity,
    isWanted: wanted,
    condition,
    card: { id: cardId, name: cardId },
    addedAt: '2026-07-01T00:00:00.000Z',
});

describe('reconcileBinderAfterTrade', () => {
    const now = '2026-07-01T00:00:00.000Z';

    test('decrements given cards and clamps at zero', () => {
        const next = reconcileBinderAfterTrade({
            entries: [binder('a', 1)],
            haveItems: [{ cardId: 'a', quantity: 3 }],
            wantItems: [],
            removeGivenFromBinder: true,
            addReceivedToBinder: false,
            now,
        });
        expect(next).toEqual([]);
    });

    test('adds received cards to binder with NM merge', () => {
        const next = reconcileBinderAfterTrade({
            entries: [binder('b', 1, { condition: 'LP' })],
            haveItems: [],
            wantItems: [{ cardId: 'b', quantity: 2, card: { _uniqueId: 'b', name: 'B' } }],
            removeGivenFromBinder: false,
            addReceivedToBinder: true,
            now,
        });
        expect(next).toHaveLength(1);
        expect(next[0].quantity).toBe(3);
        expect(next[0].condition).toBe('LP');
        expect(next[0].isWanted).toBe(false);
    });

    test('creates new NM binder row for received cards', () => {
        const next = reconcileBinderAfterTrade({
            entries: [],
            haveItems: [],
            wantItems: [{ cardId: 'c', quantity: 1, card: { _uniqueId: 'c', name: 'C' } }],
            removeGivenFromBinder: false,
            addReceivedToBinder: true,
            now,
        });
        expect(next).toHaveLength(1);
        expect(next[0]).toMatchObject({
            cardId: 'c',
            quantity: 1,
            condition: 'NM',
            isWanted: false,
        });
    });

    test('clears received cards from want list even when add is off', () => {
        const next = reconcileBinderAfterTrade({
            entries: [binder('c', 2, { wanted: true })],
            haveItems: [],
            wantItems: [{ cardId: 'c', quantity: 1 }],
            removeGivenFromBinder: false,
            addReceivedToBinder: false,
            now,
        });
        expect(next).toHaveLength(1);
        expect(next[0].quantity).toBe(1);
        expect(next[0].isWanted).toBe(true);
    });

    test('skips binder mutations when both checkboxes are off but still clears wants', () => {
        const start = [binder('g', 5), binder('r', 1, { wanted: true })];
        const next = reconcileBinderAfterTrade({
            entries: start,
            haveItems: [{ cardId: 'g', quantity: 1 }],
            wantItems: [{ cardId: 'r', quantity: 1 }],
            removeGivenFromBinder: false,
            addReceivedToBinder: false,
            now,
        });
        expect(next.find((e) => e.cardId === 'g').quantity).toBe(5);
        expect(next.find((e) => e.cardId === 'r')).toBeUndefined();
    });
});

describe('tradeLinesFromList', () => {
    test('maps uniqueId and quantity', () => {
        expect(
            tradeLinesFromList([
                { uniqueId: 'a-normal', name: 'A', quantity: 2, price: 10 },
                { name: 'No id', quantity: 1 },
            ]),
        ).toEqual([
            {
                cardId: 'a-normal',
                quantity: 2,
                card: expect.objectContaining({
                    _uniqueId: 'a-normal',
                    name: 'A',
                    marketPrice: 10,
                }),
            },
        ]);
    });
});
